src/common/meta/src/kv_backend/chroot.rs

```diff
@@ -136,121 +136,138 @@ impl KvBackend for ChrootKvBackend {
-            new_end = vec![0];
-        }
-        new_end
-    } else if !range_end.is_empty() {
-        let mut new_end = root.to_vec();
-        new_end.append(&mut range_end);
-        new_end
-    } else {
-        vec![]
+    }
+
+    fn txn_prepend_root(&self, mut txn: Txn) -> Txn {
+        let op_prepend_root = |op: TxnOp| match op {
+            TxnOp::Put(k, v) => TxnOp::Put(self.key_prepend_root(k), v),
+            TxnOp::Get(k) => TxnOp::Get(self.key_prepend_root(k)),
+            TxnOp::Delete(k) => TxnOp::Delete(self.key_prepend_root(k)),
+        };
+        txn.req.success = txn.req.success.drain(..).map(op_prepend_root).collect();
+        txn.req.failure = txn.req.failure.drain(..).map(op_prepend_root).collect();
+        txn.req.compare = txn
+            .req
+            .compare
+            .drain(..)
+            .map(|cmp| super::txn::Compare {
+                key: self.key_prepend_root(cmp.key),
+                cmp: cmp.cmp,
+                target: cmp.target,
+            })
+            .collect();
+        txn
     }
 }
 
-fn txn_prepend_root(root: &[u8], mut txn: Txn) -> Txn {
-    fn op_prepend_root(root: &[u8], op: TxnOp) -> TxnOp {
-        match op {
-            TxnOp::Put(k, v) => TxnOp::Put(key_prepend_root(root, k), v),
-            TxnOp::Get(k) => TxnOp::Get(key_prepend_root(root, k)),
-            TxnOp::Delete(k) => TxnOp::Delete(key_prepend_root(root, k)),
+#[cfg(test)]
+mod tests {
+    use std::sync::Arc;
+
+    use crate::kv_backend::chroot::ChrootKvBackend;
+    use crate::kv_backend::memory::MemoryKvBackend;
+
+    #[test]
+    fn test_prefix_key_and_range_end() {
+        fn run_test_case(pfx: &[u8], key: &[u8], end: &[u8], w_key: &[u8], w_end: &[u8]) {
+            let chroot = ChrootKvBackend::new(pfx.into(), Arc::new(MemoryKvBackend::new()));
+            assert_eq!(chroot.key_prepend_root(key.into()), w_key);
+            assert_eq!(chroot.range_end_prepend_root(end.into()), w_end);
         }
-    }
 
-    txn.req.success = txn
-        .req
-        .success
-        .drain(..)
-        .map(|op| op_prepend_root(root, op))
-        .collect();
+        // single key
+        run_test_case(b"pfx/", b"a", b"", b"pfx/a", b"");
 
-    txn.req.failure = txn
-        .req
-        .failure
-        .drain(..)
-        .map(|op| op_prepend_root(root, op))
-        .collect();
+        // range
+        run_test_case(b"pfx/", b"abc", b"def", b"pfx/abc", b"pfx/def");
 
-    txn.req.compare = txn
-        .req
-        .compare
-        .drain(..)
-        .map(|cmp| super::txn::Compare {
-            key: key_prepend_root(root, cmp.key),
-            cmp: cmp.cmp,
-            target: cmp.target,
-        })
-        .collect();
+        // one-sided range (HACK - b'/' + 1 = b'0')
+        run_test_case(b"pfx/", b"abc", b"\0", b"pfx/abc", b"pfx0");
 
-    txn
+        // one-sided range, end of keyspace
+        run_test_case(b"\xFF\xFF", b"abc", b"\0", b"\xff\xffabc", b"\0");
```

Comment:
Nit: why is the `\xFF\xFF` in upper case?

Reply:
If you like consistent to lowercase, just change to it. It's not technically different.